### PR TITLE
Elasticsearch ^8.0 support

### PR DESCRIPTION
**Pull request by sporchia**
This pull request contains breaking changes.

Elasticsearch 8.x introduced many breaking changes to the client. It seems to have been a complete overhaul.

8.x also introduced the native Client from ClientBuilder so there is no longer a need for the following files:

src/Infrastructure/Elastic/ElasticClientFactory.php - removed
src/Infrastructure/Elastic/ElasticClientBuilder.php- removed
and associated tests:

tests/Unit/ElasticClientBuilderTest.php - removed
tests/Unit/ElasticClientFactoryTest.php - removed
Client can now be injected via native ClientBuilder so ElasticClientBuilder was replaced with Elastic's native ClientBuilder and documentation has been updated to reflect this.

Also this updates tests to match how Elasticsearch suggests you mock their requests (using Guzzle in this case, see changes in FinderTest.php)

this is an expansion on #239 getting all the CI to work properly

### CODE REVIEW

**Comment by sporchia** [comment]
Been testing this locally and seems to be working fine, any possible traction on merging?

**Comment by drmmr763** [comment]
@Jeroen-G please?

**Comment by Jeroen-G** [comment]
I'm in the middle of renovating so time is limited but here are a few of my thoughts. Sorry if I missed something, I haven't read the ES changelog (and currently no project that requires it😬). But of course happy to support the future! 🚀

Is this really the minimum that is necessary? Quite some breaking changes indeed.

There are a few like I'd like differently, like config keys named like api_key, but that can be done post-merge.

@blackshadev I would also appreciate your review on this.

@drmmr763 are you able to functionally test this?

If anyone has some extra time to open a PR to update the [demo repo](https://github.com/Jeroen-G/explorer-demo) that would be awesome 🙏

**Comment by blackshadev** [comment]
@sporchia Sorry for the wait! This wasn't on our radar at all.

I am agreeing with Jeroen here, I would like for the config to be backwards compatible with our old config. Why Do we need to change to the Elastic one. It has casing on weird places, totally not inline with laravel's own config keys.

File: src/Application/IndexAdapterInterface.php [hunk 1]
```diff
@@ -16,7 +17,7 @@ public function delete(IndexConfigurationInterface $indexConfiguration): void;
 
     public function flush(string $index): void;
 
-    public function createNewWriteIndex(IndexConfigurationInterface $indexConfiguration): string;
+    public function createNewWriteIndex(AliasedIndexConfiguration $indexConfiguration): string;
```
Comment (blackshadev) [inline src/Application/IndexAdapterInterface.php:20]: Just a question, why this change? Does this make something easier? I would rather have interfaces depend on other interfaces. Even more so because it already is dependent on the `IndexConfigurationInterface`

File: src/Application/Results.php [hunk 2]
```diff
@@ -5,14 +5,15 @@
...
 
 class Results implements Countable
 {
     private array $rawResults;
 
-    public function __construct(array $rawResults)
+    public function __construct(array|Elasticsearch $rawResults)
     {
-        $this->rawResults = $rawResults;
+        $this->rawResults = is_array($rawResults) ? $rawResults : $rawResults->asArray();
```
Comment (blackshadev) [inline src/Application/Results.php:16]: I this needed? Isn't it nicer if you call the `Elasticsearch::asArray` on the calling site?

File: docs/connection.md [hunk 3]
```diff
@@ -101,12 +101,14 @@ To disable TLS verification set it to `false`. **NOT recommended for production*
...
-                'cert' => ['path/to/cert.pem', 'passphrase'],
-                'key' => ['path/to/key.pem', 'passphrase'],
+            'Hosts' => ['localhost:9200'],
+            'SSLCert' => [
+                'cert' => './path/to/cert.crt',
+                'password' => null,
+            ],
+            'SSLKey' => [
+                'key' => './path/to/key.key',
+                password' => null,
```
Comment (blackshadev) [inline docs/connection.md:111]: I really think we should be backwards compatible with the old config, I rather not ask our users to rewrite their configs after an update

**Review by blackshadev — CHANGES_REQUESTED** [review]
If you do not want to make these changes yourself I can consult with JeroenG what he finds preferable and otherwise try to create an PR with you code and the requested changes, it will take some time though.

**Comment by drmmr763** [comment]
@Jeroen-G I forced my composer.json to pull from @sporchia's repo earlier today and I've been able to run all the operations I would expect without any issue. Re-indexed all my models and various customizations I need are there. 

In regard to the latest version - there are some pretty important things (to me) that this latest version of the client supports that aren't in 7.x Specifically inference pipelines:

https://www.elastic.co/guide/en/elasticsearch/reference/current/semantic-search-inference.html

So far, everything has worked as expected. I will say, I do not use the full scope of Laravel Scout. My frontend app calls Elastic search directly by-passing Laravel entirely. I only use scout for maintaining my elastic index.

File: src/Application/IndexAdapterInterface.php (hunk 1 above)
Comment (sporchia) [reply src/Application/IndexAdapterInterface.php:20]: gonna take a moment for me to get my mind back into it, but I'm pretty sure this change because this doesn't actually work with anything other than the `AliasedIndexConfiguration` it's expected in places further down the code, so allowing an interface, but needing the real class is weird, as it just needs the real class. but it could be put back, it would just be a vector for unknown errors

File: src/Application/Results.php (hunk 2 above)
Comment (sporchia) [reply src/Application/Results.php:16]: I believe this might be going back to the original PR i based this on, happy to make adjustments here as I'm not super tied to anything, we have been using the branch in our production for a few months now. This was just what was required to assure that the tests would pass and the code as originally modified would run.

File: docs/connection.md (hunk 3 above)
Comment (sporchia) [reply docs/connection.md:111]: I'll look into it and adjust, I believe this was also a remnant from the original PR that didn't have passing tests.

**Comment by blackshadev** [comment]
I just read the changelog and this PR has another big problem: https://github.com/elastic/elasticsearch-php?tab=readme-ov-file#compatibility . If we install v8 of elasticsearch, v7 will not work.

So we would either:
- Need to maintain 2 separate versions, as elasticsearch does
- Ensure we pick the correct version of the elasticsearch client based on "things" and rework it such that it is transparent. 

I find it really annoying, but we probably should go for the first option, as the second will introduce WAAAAAAY more code and how in the world would we switch between the v7 or v8 dependency.

File: src/Application/IndexAdapterInterface.php (hunk 1 above)
Comment (blackshadev) [reply src/Application/IndexAdapterInterface.php:20]: I do not follow your reasoning. I know currently `createNewWriteIndex` is only called internally and only passed the class `AliasedIndexConfiguration`. But internally we do not rely on the fact that `AliasedIndexConfiguration` is passed. That is why we have the interface, that the contract we should (and do) adher to.  And which php will enforce for us do you strict typing. To me this isn't a vector of unknown errors. It is an a stricter contract between the code. I do also do not see why it should break with `IndexConfigurationInterface` as it has always worked any non of your changes breaks the original contract.

If you could explain more about the " vector for unknown errors" you see I could maybe lean your way

**Comment by drmmr763** [comment]
Just noticed in testing this that I think the logging is broken. Not sure if its because of the config changes or what. Looking into it more.

**Comment by drmmr763** [comment]
Just some experience I've had testing this in production. It seems the default client builder provided by Elastic uses a `no resurrect` node pool selector:

https://www.elastic.co/guide/en/elasticsearch/client/php-api/current/node_pool.html

This has the unfortunate consequence of an intermittent connection issue taking the node offline permanently in the node pool and never bringing it back. 

from their docs:
```
The NoResurrect option does not try to resurrect the node that has been marked as dead. If you want, you can specify the ElasticsearchResurrect class to check if a node that was dead is online again (resurrected).
```

To get around this issue I am implementing my own service provider to add in the Resurrect class as advised. But, imho this default behavior is not very fun and I would advise maybe re-writing the default provider to use the resurrect class from the get-go.

**Comment by Jeroen-G** [comment]
Closed in favour or #253
